salt/transport/tcp.py

```diff
@@ -773,38 +812,200 @@ def handle_future(future):
+            if self.presence_events:
+                data = {'new': [id_],
+                        'lost': []}
+                self.event.fire_event(
+                    data,
+                    salt.utils.event.tagify('change', 'presence')
+                )
+                data = {'present': list(self.present.keys())}
+                self.event.fire_event(
+                    data,
+                    salt.utils.event.tagify('present', 'presence')
+                )
+
+    def _remove_client_present(self, client):
+        id_ = client.id_
+        if id_ is None or id_ not in self.present:
+            return
+
+        clients = self.present[id_]
+        if client not in clients:
+            return
+
+        clients.remove(client)
+        if len(clients) == 0:
+            del self.present[id_]
+            if self.presence_events:
+                data = {'new': [],
+                        'lost': [id_]}
+                self.event.fire_event(
+                    data,
+                    salt.utils.event.tagify('change', 'presence')
+                )
+                data = {'present': list(self.present.keys())}
+                self.event.fire_event(
+                    data,
+                    salt.utils.event.tagify('present', 'presence')
+                )
+
+    @tornado.gen.coroutine
+    def _stream_read(self, client):
+        unpacker = msgpack.Unpacker()
+        while not self._closing:
+            try:
+                client._read_until_future = client.stream.read_bytes(4096, partial=True)
+                wire_bytes = yield client._read_until_future
+                unpacker.feed(wire_bytes)
+                for framed_msg in unpacker:
+                    body = framed_msg['body']
+                    if body['enc'] != 'aes':
+                        # We only accept 'aes' encoded messages for 'id'
+                        continue
+                    crypticle = salt.crypt.Crypticle(self.opts, salt.master.SMaster.secrets['aes']['secret'].value)
+                    load = crypticle.loads(body['load'])
+                    if not self.aes_funcs.verify_minion(load['id'], load['tok']):
+                        continue
+                    client.id_ = load['id']
+                    self._add_client_present(client)
+            except tornado.iostream.StreamClosedError as e:
+                log.debug('tcp stream to {0} closed, unable to recv'.format(client.address))
+                client.close()
+                self._remove_client_present(client)
+                self.clients.discard(client)
+                return
+            except Exception as e:
+                log.error('Exception parsing response', exc_info=True)
+                continue
 
     def handle_stream(self, stream, address):
         log.trace('Subscriber at {0} connected'.format(address))
-        self.clients.append((stream, address))
+        client = PubServerClient(stream, address)
```

Comment:
We might want to rename `PubServerClient`, only because that class makes me think its the client to the pub server-- to be used on a minion. When in reality it for encapsulating a client of the pubserver. Maybe we call it `Subscriber`?